ffmpeg_tools/commands.py

```diff
@@ -180,6 +236,67 @@ def merge_videos_command(input_file, output):
     return cmd, input_file
 
 
+def replace_streams(input_file,
+                    replacement_source,
+                    output_file,
+                    stream_type):
+
+    assert os.path.isfile(input_file)
+    assert os.path.isfile(replacement_source)
+    assert not os.path.exists(output_file)
+
+    cmd = replace_streams_command(
+        input_file,
+        replacement_source,
+        output_file,
+        stream_type)
+
+    exec_cmd(cmd)
+
+
+def replace_streams_command(input_file,
+                            replacement_source,
+                            output_file,
+                            stream_type):
+    """
+    Builds a ffmpeg command that can be used to create a new video file with
+    all streams of a specific type replaced with streams of the same type from
+    another video file.
+
+    :param input_file: Container from which all the streams of types other than
+        the specified one will be taken. Must exist.
+    :param replacement_source: Container from which streams of the specified
+        type will be taken. Must exist.
+    :param output_file: Container to put the streams in. Must not exist.
+    :param stream_type: Stream type specifier.
+        See https://ffmpeg.org/ffmpeg.html#Stream-specifiers.
+        The following values are supported:
+            - `v` - same as `V`.
+            - `V` - video streams which are not attached pictures, video
+                    thumbnails or cover arts.
+            - `a` - audio streams.
+            - `s` - subtitle streams.
+            - `d` - data streams.
+            - `t` - attachments.
+    """
+    assert stream_type in ['v', 'V', 'a', 's', 'd', 't']
```

Comment:
This probably should be exception if it can happen in production code.

Reply:
It depends. Is `ffmpeg_tools` supposed to validate and sanitize its input or is it the responsibility of the caller?

Currently these functions do not validate anything - parameter values are passed straight to ffmpeg with minimal processing. So I did not add any validations for the new functions either. My assumption was that they are (or should be) on a different layer, closer to the user. E.g. at the task level.

This `assert` was not added for validation. It's here to make sure that it's easy to notice at development time if you happen to use the function incorrectly. ffmpeg's `-map` supports a wide range of stream selectors and if you use something weird the function is likely to just do something wrong instead of outright failing. The assertion stops it immediately and fails loudly instead of letting it continue. It also makes it clear that the failure is not a problem in the implementation but rather in the call itself - the function explicitly supports only these specific values.

An exception would serve that purpose too. I simply think that `assert` is enough when it's just a sanity check rather than a validation.

Reply:
Ok but there's no validation for this in ffmpeg_task either. You can pass whatever you want to docker container parameters. Better make it clean and turn it to exception.

You are right that we didn't make enough validation in many places, but ffmpeg-tools has many utilities to validate parameters and they should be used, after you update supported containers/codecs lists.

Docker container should handle all input parameters properly. Check all other asserts if they shouldn't be replaced with exceptions too.

Reply:
Done. I replaced this assertion with a parameter validation for this single case.

I've defined an exception called `InvalidArguement` for the error. There's no separate module for exceptions and it does not belong in `validations.py` in my opinion so I left it in `commands.py`.

As for other asserts - the only ones still present are ones that check whether the input file exists  and output file does not. This is mostly so that we can catch situations where a file would be overwritten. And that's only because currently there is no situation where we want to overwrite it. But this could change and overwriting it is not necessarily an error. It might be a feature if you do want it. So I'm not sure if we want the commands to actually validate that the file exists or not. I'd personally leave it as asserts for now.